File: system-base/src/main/java/com/levnovikov/system_base/Interactor.java
package com.levnovikov.system_base;

import android.os.Parcelable;

import com.levnovikov.system_base.state.ActivityState;
import com.levnovikov.system_base.state.NodeState;

public abstract class Interactor<R extends Router> {

    protected R router;
    protected ActivityState activityState;

    public Interactor(R router, ActivityState activityState) {
        this.router = router;
        this.activityState = activityState;
    }

    public void onGetActive() {
        final NodeState state = getNodeState();
        if (state != null) {
            router.setState(state);
        }
    }

    private NodeState getNodeState() {
        return activityState.findNodeState(router.getClass());
    }

    protected Parcelable getNodeStateData() {
        final NodeState state = getNodeState();
        return state != null ? state.data() : null;
    }

    protected boolean hasSavedState() {
        return activityState.findNodeState(router.getClass()) != null;
    }
}

File: system-base/src/main/java/com/levnovikov/system_base/Interactor.kt
package com.levnovikov.system_base

import android.os.Parcelable

import com.levnovikov.system_base.node_state.ActivityState
import com.levnovikov.system_base.node_state.NodeState

abstract class Interactor<R : Router>(
        protected var router: R,
        protected var activityState: ActivityState) {

    protected val nodeState: NodeState? = activityState.findNodeState(router.javaClass)

    protected val nodeStateData: Parcelable?
        get() {
            val state = nodeState
            return state?.data
        }

    open fun onGetActive() {
        val state = nodeState
        if (state != null) {
            router.setState(state)
        }
    }

    protected fun hasSavedState(): Boolean {
        return activityState.findNodeState(router.javaClass) != null
    }
}
